src/passes/TypeGeneralizing.cpp

```diff
@@ -323,81 +394,469 @@ struct TransferFn : OverriddenVisitor<TransferFn> {
+    push(wasm.getTable(curr->table)->type);
+  }
+
+  void visitTableCopy(TableCopy* curr) {
+    // Cannot generalize table types yet.
+  }
+
   void visitTry(Try* curr) { WASM_UNREACHABLE("TODO"); }
   void visitThrow(Throw* curr) { WASM_UNREACHABLE("TODO"); }
   void visitRethrow(Rethrow* curr) { WASM_UNREACHABLE("TODO"); }
   void visitTupleMake(TupleMake* curr) { WASM_UNREACHABLE("TODO"); }
   void visitTupleExtract(TupleExtract* curr) { WASM_UNREACHABLE("TODO"); }
+
   void visitRefI31(RefI31* curr) { pop(); }
   void visitI31Get(I31Get* curr) { push(Type(HeapType::i31, Nullable)); }
-  void visitCallRef(CallRef* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitRefTest(RefTest* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitRefCast(RefCast* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitBrOn(BrOn* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitStructNew(StructNew* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitStructGet(StructGet* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitStructSet(StructSet* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayNew(ArrayNew* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayNewData(ArrayNewData* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayNewElem(ArrayNewElem* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayNewFixed(ArrayNewFixed* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayGet(ArrayGet* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArraySet(ArraySet* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayLen(ArrayLen* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayCopy(ArrayCopy* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayFill(ArrayFill* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayInitData(ArrayInitData* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitArrayInitElem(ArrayInitElem* curr) { WASM_UNREACHABLE("TODO"); }
-  void visitRefAs(RefAs* curr) { WASM_UNREACHABLE("TODO"); }
+
+  void visitCallRef(CallRef* curr) {
+    auto sigType = curr->target->type.getHeapType();
+    if (sigType.isBottom()) {
+      // This will be emitted as an unreachable, so impose no requirements on
+      // the arguments, but do require that the target continue to have bottom
+      // type.
+      clearStack();
+      push(Type(HeapType::nofunc, Nullable));
+      return;
+    }
+
+    auto sig = sigType.getSignature();
+    auto numParams = sig.params.size();
+    std::optional<Type> resultReq;
+    if (sig.results.isRef()) {
+      resultReq = pop();
+    }
+
+    // We have a choice here: We can either try to generalize the type of the
+    // incoming function reference or the type of the incoming function
+    // arguments. Because function parameters are contravariant, generalizing
+    // the function type inhibits generalizing the arguments and vice versa.
+    // Attempt to split the difference by generalizing the function type only as
+    // much as we can without imposing stronger requirements on the arguments.
```

Comment:
A little like RefCast, this seems somewhat speculative to me. Picking one of the two other options would be simpler. I'd pick function parameters myself. However, unlike RefCast I don't have an intuition here about whether the complexity is worth it or not, or even between the options.

Reply:
I don't think the differences in code complexity between the options are too large in the grand scheme of things. Let's leave this as-is and make sure to experiment with different choices on real-world programs.

Reply:
Sounds good.